metagraph/src/common/vectors/wavelet_tree.cpp

```diff
@@ -458,6 +483,13 @@ TAlphabet partite_vector<t_bv>::operator[](uint64_t i) const {
     return int_vector_[i];
 }
 
+template <class t_bv>
+std::pair<uint64_t, TAlphabet> partite_vector<t_bv>::inverse_select(uint64_t i) const {
+    assert(i < size());
+    TAlphabet val = int_vector_[i];
+    return std::make_pair(rank(val, i), val);
```

Comment:
so for this type, the old implementation makes more sense. Can you check it inside `fwd(edge)` and avoid `inverse_select` for `partite_vector`?

Reply:
The extra dynamic_cast for this check slows things down.
```
volatile auto t = dynamic_cast<const partite_vector<>*>(W_);
    std::ignore = t;
    // if (!dynamic_cast<const partite_vector<>*>(W_)) {
        auto [rank, c] = W_->inverse_select(i);
        assert(i == 1 || c != kSentinelCode);
        if (c == kSentinelCode) {
            --rank;
        } else if (c >= alph_size) {
            // this branch is rare on graphs with low average in-degree
            c %= alph_size;
            rank = rank_W(i, c);
        }
        // get the index of the target node
        node_index target_node = NF_[c] + rank;
        // get the index of the representative edge with that target node
        return select_last(target_node);
    // }

    // return fwd(i, get_W(i) % alph_size);
```
gets me this on the stat representation
```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
BM_BOSS_get_W_and_fwd         0.338 us        0.337 us      2094306
BM_BOSS_no_get_W_and_fwd      0.532 us        0.531 us      1154902
```

whereas if I don't do the cast at all
```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
BM_BOSS_get_W_and_fwd         0.386 us        0.385 us      1904289
BM_BOSS_no_get_W_and_fwd      0.379 us        0.378 us      1827670
```

And I'm not sure if it's even worth doing all this, since this PR doesn't seem to be negatively affecting fwd on partite_vector anyway
```
-------------------------------------------------------------------
Benchmark                         Time             CPU   Iterations
-------------------------------------------------------------------
BM_BOSS_get_W_and_fwd         0.443 us        0.442 us      1511411
BM_BOSS_no_get_W_and_fwd      0.420 us        0.419 us      1681473
```

Reply:
An alternative would be to do the dynamic cast one in the BOSS constructor and loader and keep using that.